src/Essentials/src/FileSystem/FileSystem.ios.cs

```diff
@@ -158,68 +160,134 @@ protected override void Dispose(bool disposing)
-
-			FullPath = Guid.NewGuid().ToString() + FileExtensions.Png;
-			FileName = FullPath;
+			this.img = img;
+			this.metadata = metadata;
+			NameWithoutExtension = name;
+#pragma warning disable CA1422
+			ContentType = GetMIMEType(OldUTType.JPEG);
+			Extension = GetExtension(OldUTType.JPEG);
+#pragma warning restore CA1422
+			Type = GetFileType(ContentType);
+			FileName = GetFileName(NameWithoutExtension, Extension);
 		}
 
 		internal override Task<Stream> PlatformOpenReadAsync()
 		{
-			data ??= uiImage.AsPNG();
+			imgWithMetadata ??= GetImageWithMeta();
+			return Task.FromResult(imgWithMetadata?.AsStream());
+		}
 
-			return Task.FromResult(data.AsStream());
+		public NSMutableData GetImageWithMeta()
+		{
+			if (img == null || metadata == null)
+				return null;
+
+			using var source = CGImageSource.FromData(img.AsJPEG());
+			var destData = new NSMutableData();
+			using var destination = CGImageDestination.Create(destData, source.TypeIdentifier, 1, null);
+			destination.AddImage(source, 0, metadata);
+			destination.Close();
+			DisposeSources();
+			return destData;
+		}
+
+		protected internal override void PlatformDispose()
+		{
+			imgWithMetadata?.Dispose();
+			imgWithMetadata = null;
+			DisposeSources();
+			base.PlatformDispose();
+		}
+
+		void DisposeSources()
+		{
+			img?.Dispose();
+			img = null;
+			metadata?.Dispose();
+			metadata = null;
 		}
 	}
 
-	class PHAssetFileResult : FileResult
+	class PHPickerFileResult : MediaFileResult
 	{
-		readonly PHAsset phAsset;
+		readonly string identifier;
+		NSItemProvider provider;
 
-		internal PHAssetFileResult(NSUrl url, PHAsset asset, string originalFilename)
-			: base()
+		internal PHPickerFileResult(NSItemProvider provider)
 		{
-			phAsset = asset;
+			this.provider = provider;
+			NameWithoutExtension = provider?.SuggestedName;
+
+			identifier = GetIdentifier(provider?.RegisteredTypeIdentifiers);
 
-			FullPath = url?.AbsoluteString;
-			FileName = originalFilename;
+			if (string.IsNullOrWhiteSpace(identifier))
+				return;
+
+			Extension = GetExtension(identifier);
+			ContentType = GetMIMEType(identifier);
+			Type = GetFileType(ContentType);
+			FileName = GetFileName(NameWithoutExtension, Extension);
 		}
 
-		[System.Runtime.Versioning.UnsupportedOSPlatform("ios13.0")]
-		internal override Task<Stream> PlatformOpenReadAsync()
-		{
-			var tcsStream = new TaskCompletionSource<Stream>();
+		internal override async Task<Stream> PlatformOpenReadAsync()
+			=> (await provider?.LoadDataRepresentationAsync(identifier))?.AsStream();
 
-			PHImageManager.DefaultManager.RequestImageData(phAsset, null, new PHImageDataHandler((data, str, orientation, dict) =>
-				tcsStream.TrySetResult(data.AsStream())));
+		protected internal override void PlatformDispose()
+		{
+			provider?.Dispose();
+			provider = null;
+			base.PlatformDispose();
+		}
 
-			return tcsStream.Task;
+		private string GetIdentifier(string[] identifiers)
+		{
+			if (!(identifiers?.Length > 0))
```

Comment:
Making thins complicated. We can check for null, in the later part of the method you do FirstOrDefault, so if the array is empty, you will get the default => null.